docs/architecture/adr-007-light-client-supervisor-ergonomics.md

```diff
@@ -0,0 +1,53 @@
+# ADR 007: Light Client Supervisor Ergonomics
+
+## Changelog
+
+* 2020-06-30: Initial draft
+
+## Context
+
+The initial approach to use callbacks in order to capture async responses
+from requests issued by the `Handle` to its `Supervisor` introduces a couple of
+has some drawbacks, i.e. try semantics, scope and borrow complexities. As the
+general pattern seems to be that a bounded channel is constructed to transport
+the response it could be reworked to be the actual value send with the event to
+the Supervisor. Which in turn send back the response on the typed channel to the
+Handle. As an added side-effect this could remove the need for the Callback
+abstraction all together and fully embraces CSP style concurrency.
```

Comment:
I can't think of any drawback removing the Callback abstractions. @brapse @romac We aren't missing any larger context here, right? (e.g. the callbacks weren't supposed to later carry some traceability function, like a logger etc? and either way, this can equivalently be modelled using the channel / csp abstraction).

Reply:
> We aren't missing any larger context here, right? (e.g. the callbacks weren't supposed to later carry some traceability function, like a logger etc?

No, the callbacks were just meant as a thin abstraction over the underlying channels, that's it.

> and either way, this can equivalently be modelled using the channel / csp abstraction).

Yep, I concur in https://github.com/informalsystems/tendermint-rs/pull/403#discussion_r448474189